gcc/d/decl.cc

```diff
@@ -1236,17 +1252,25 @@ get_symbol_decl (Declaration *decl)
 tree
 declare_extern_var (tree ident, tree type)
 {
+  /* If the VAR_DECL has already been declared, return it.  */
+  if (IDENTIFIER_DECL_TREE (ident))
+    return IDENTIFIER_DECL_TREE (ident);
+
   tree name = IDENTIFIER_PRETTY_NAME (ident)
     ? IDENTIFIER_PRETTY_NAME (ident) : ident;
   tree decl = build_decl (input_location, VAR_DECL, name, type);
 
+  IDENTIFIER_DECL_TREE (ident) = decl;
   d_keep (decl);
 
   SET_DECL_ASSEMBLER_NAME (decl, ident);
   DECL_ARTIFICIAL (decl) = 1;
   TREE_STATIC (decl) = 1;
   TREE_PUBLIC (decl) = 1;
 
+  /* Mark it needed so we don't forget to emit it.  */
+  mark_needed (decl);
```

Comment:
Always emit compiler generated variables (this is used by vtbl, init, typeinfo, classinfo).  This could be improved, but not sure how as generally this is always needed as you can never be sure who might have an extern reference to it until link-time.